src/proof/verify.rs

```diff
@@ -0,0 +1,217 @@
+//! Proof verification logic.
+
+use crate::{
+    nodes::{rlp_node, word_rlp, TrieNode, CHILD_INDEX_RANGE},
+    proof::ProofVerificationError,
+    EMPTY_ROOT_HASH,
+};
+use alloc::vec::Vec;
+use alloy_primitives::{Bytes, B256};
+use alloy_rlp::Decodable;
+use nybbles::Nibbles;
+
+/// Verify the proof for given key value pair against the provided state root.
+/// Returns the leaf node value for the given key.
+pub fn verify_proof<'a, I>(
+    proof: I,
+    root: B256,
+    key: B256,
+) -> Result<Option<Vec<u8>>, ProofVerificationError>
+where
+    I: IntoIterator<Item = &'a Bytes>,
+{
+    let mut proof = proof.into_iter().peekable();
+
+    if proof.peek().is_none() {
+        return if root == EMPTY_ROOT_HASH {
+            Ok(None)
+        } else {
+            return Err(ProofVerificationError::RootMismatch {
+                got: EMPTY_ROOT_HASH,
+                expected: root,
+            });
+        };
+    }
+
+    let target = Nibbles::unpack(key);
+    let mut walked_path = Nibbles::default();
+    let mut expected_value = Some(word_rlp(&root));
+    for node in proof {
+        if Some(rlp_node(node)) != expected_value {
+            let got = Some(Bytes::copy_from_slice(node));
+            let expected = expected_value.map(|b| Bytes::copy_from_slice(&b));
+            return Err(ProofVerificationError::ValueMismatch { path: walked_path, got, expected });
+        }
+
+        expected_value = match TrieNode::decode(&mut &node[..])? {
+            TrieNode::Branch(branch) => 'val: {
+                if let Some(next) = target.get(walked_path.len()) {
+                    let mut stack_ptr = branch.as_ref().first_child_index();
```

Comment:
TODO: The Branch Node's stack may have an arbitrary length whereas it should equal to the number of 1s in the state mask